Add action to copy index metadata when reindexing data stream indices

parkertimmins opened the pull request:
When reindexing data stream indices, parts of the index metadata needs to be copied from the source index to destination index, so that ILM and Data stream lifecycle continue to function properly. This adds a new `CopyLifecycleIndexMetadataTransportAction` which copies the following metadata:
* creation date 
* rollover info
* ILM custom metadata

parkertimmins committed:
Setup index metadata update action skeleton

parkertimmins committed:
copy index metadata from source to dest

parkertimmins committed:
Set ilm and data stream lifecycle on dest index metadata

parkertimmins committed:
Add create date, call from reindex

parkertimmins committed:
Add IT tests

parkertimmins committed:
Merge branch 'main' into reindex-data-stream-copy-index-metadata

parkertimmins committed:
Add test for creation date

parkertimmins committed:
Merge branch 'main' into reindex-data-stream-copy-index-metadata

parkertimmins committed:
rollover info test

parkertimmins committed:
trying add ilm tests

parkertimmins committed:
Merge branch 'main' into reindex-data-stream-copy-index-metadata

parkertimmins committed:
add ilm tests

Commit:
[CI] Auto commit changes from spotless

masseyke commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/action/ReindexDataStreamIndexTransportAction.java:
```diff
@@ -334,6 +335,24 @@ private void copyOldSourceSettingsToDest(
+        ActionListener<AcknowledgedResponse> listener,
+        TaskId parentTaskId
+    ) {
+        logger.debug("Updating settings on destination index after reindex completes");
```
Need to fix this log message

parkertimmins committed:
Fix ILM state tests

parkertimmins committed:
Merge branch 'main' into reindex-data-stream-copy-index-metadata

parkertimmins committed:
unneeded code

masseyke commented on x-pack/plugin/migrate/src/internalClusterTest/java/org/elasticsearch/xpack/migrate/action/ReindexDataStreamTransportActionIT.java:
```diff
@@ -26,6 +26,7 @@
 import org.elasticsearch.test.ESIntegTestCase;
 import org.elasticsearch.transport.TransportService;
 import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.LocalStateCompositeXPackPlugin;
```
unnecessary import I think

Commit:
[CI] Auto commit changes from spotless

parkertimmins committed:
Fixes from review:

- add action to internal user for reindex data stream
- make request an IndicesRequest
- rename to CopyLifecycleIndexMetadataAction
- fix ILM state test

parkertimmins commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -2077,6 +2077,12 @@ public Builder putRolloverInfo(RolloverInfo rolloverInfo) {
         }
 
+        public Builder putRolloverInfos(Map<String, RolloverInfo> rolloverInfos) {
+            this.rolloverInfos.clear();
```
clearing the existing infos is a bit weirds, but seems better than making rolloverInfos non-final

parkertimmins committed:
Fix DataStreamsUpgradeIT ILM test

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

parkertimmins committed:
Remove references to reindex data stream feature flag

Commit:
[CI] Auto commit changes from spotless

masseyke reviewed (approved):
LGTM

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [9.0](https://github.com/elastic/elasticsearch/pull/122555)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/122555">](https://github.com/elastic/elasticsearch/pull/122555) |
| ❌ |  8.18  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 122535`

parkertimmins commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/122559">](https://github.com/elastic/elasticsearch/pull/122559)|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/122561">](https://github.com/elastic/elasticsearch/pull/122561)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.2 -->
